### PR TITLE
[4.0] Markdown content entry

### PR DESCRIPTION
**Pull request by brianteeman**
### Summary of Changes
Adds the ability to use markdown style content editing with tinymce
This is disabled by default.

### Testing Instructions
Apply the PR
In the tinymce plugin select SET 0
Scroll down and turn ON the markdown style editing
Save and close and then open an article

Insert standard markdown. As soon as you press space or enter the content will be rendered as html. It is always saved as html.

#### Supported markdown
```
# This is an <h1> tag
## This is an <h2> tag
### This is an <h3> tag
#### This is an <h4> tag
##### This is an <h5> tag
###### This is an <h6> tag
*This text will be italic*
**This text will be bold**
1. ordered list item
* unordered list item
- unordered list item
```

#### Editor
mark

#### Plugin
image

### Documentation Changes Required
yes

### CODE REVIEW

**Comment by ChristineWk** [comment]
I have tested this item :white_check_mark: successfully on 9ee149b3e151a201c269768bcf1d5bc683d1abcb<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31911">issues.joomla.org/tracker/joomla-cms/31911</a>.</sub>

**Comment by toivo** [comment]
I have tested this item :white_check_mark: successfully on 9ee149b3e151a201c269768bcf1d5bc683d1abcb<br /><br />Tested successfully in Beta7-dev of 9 January.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31911">issues.joomla.org/tracker/joomla-cms/31911</a>.</sub>

**Comment by alikon** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31911">issues.joomla.org/tracker/joomla-cms/31911</a>.</sub>

**Comment by infograf768** [comment]
imho his needs a description with example as the label is not enough to understand what this field means.

**Comment by infograf768** [comment]
Also, it is a new feature, and therefore not for 4.0

To understand fully what it meant, I had to go to https://fr.wikipedia.org/wiki/Markdown (for French details)

**Comment by chmst** [comment]
Good feature. But only few users will know what it means and how to use it. I agree with @infograf768 that this needs good description.

**Comment by infograf768** [comment]
Also, one has to understand that if the field is set to yes, then when one wants to write a specific markdown code NOT as a markdown, he/she will get weird results
For example if I use # and I do NOT want a h1. 

This means that the description should include a link to a page like the one in wikipedia, which has equivalent pages for most languages so that Translators can add custom link.

**Comment by infograf768** [comment]
Something like
`PLG_TINY_FIELD_TEXTPATTERN_DESC="Uses \"Markdown\" to apply styles. This may have undesired consequences when using its specific characters. Full details can be found here <a href=\"https://en.wikipedia.org/wiki/Markdown" target=\"blank\">Markdown codes</a>"`

After discussion in Maintainers chat, setting this back to pending.

**Comment by infograf768** [comment]
Pending<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31911">issues.joomla.org/tracker/joomla-cms/31911</a>.</sub>

**Comment by infograf768** [comment]
In fact wikipedia, although explaining what is markdown, does not explain which patterns can be used
Some explanation here, but not a full list of default possibilities.
https://www.tiny.cloud/docs/plugins/opensource/textpattern/

In English only, alas.

**Comment by brianteeman** [comment]
As stated in the original post this needs documentation which I will write

As for the undesired consequences - thats why its not enabled by default although its unlikely to cause any problems

As for it being a new feature. Its not really as we are already shipping the code I have just enabled the ability to use it. Or in other words fixed a bug that new features in tinymce were not considered when this new version was introduced in joomla 4

While this might be a really minor change its the sort of thing that appeals to users and is a more marketable feature than most oif the changes in j4

**Comment by alikon** [comment]
if everything is considered a new feature....and have to wait till 4.1, while we are still in beta...
:-1:

**Comment by infograf768** [comment]
> As stated in the original post this needs documentation which I will write

This is not enough. We need a description to alert the user about the feature, as stated above.

**Comment by brianteeman** [comment]
image

**Comment by infograf768** [comment]
Sorry, for me it is not enough... If I had not seen this PR and looked for Markdown on the internet I would have no idea what it is for. And, as I wrote, it should be an alert to the user that this may interact with usual TinyMCE behaviour in case some of this markup syntax is undesired.

**Comment by brianteeman** [comment]
I guess that you have never used markdown on github or the issue tracker then. I can't fix that.

**Comment by infograf768** [comment]
Also, at reading the wikipedia article, I looks like `Markdown` should always be capitalized.
https://en.wikipedia.org/wiki/Markdown

**Comment by brianteeman** [comment]
No point in me changing that as you will still object to moving into the 21st century

**Comment by infograf768** [comment]
> I guess that you have never used markdown on github or the issue tracker then. I can't fix that.

I use some of them but I did not even know the name of that functionality. 
Anyway, it is totally irrelevant to the point at stake as it is certainly not a common knowledge among J! user.

> No point in me changing that as you will still object to moving into the 21st century

That is an argument you and some others repeat all the time when someone does not agree, and it is frankly insulting.
Did I say we should not include this feature in J ? 
Certainly not. I just said users should be informed of what they do to make J friendly. 
AND NOT ALWAYS USE THE PRETEXT OF DOCS to not do so.
Your refusal to even consider this is just shocking.

**Comment by brianteeman** [comment]
> Anyway, it is totally irrelevant to the point at stake as it is certainly not a common knowledge among J! user.

That's a fact because you know the mind of all joomla users??

Did you also insist the same for the codemirror editor?
image

Seriously JM you have to stop insisting that because you don't know about something then no one else possibly can know.

This is an optional feature that is disabled by default. If you don't know what it does then a normal user would either ignore it or check the help (which I already said I would update). That's the same for everything

**Comment by ChristineWk** [comment]
> Good feature. But only few users will know what it means and how to use it. 

Exactly. 

Peace **please** :-)
A few words from a normal user like me. Not a programmer or anything else.

Never used Markdown before. A few months ago, I asked someone what this was for. Then got information that it is used here on Github. Then I saw this PR and tried it out. 

The instructions for this were helpful to me as well as the gif. Otherwise I would **not** have known how to get to this result. For me this PR was to test the function. Of course, I can't say whether I'll ever use it in conjunction with Editor.

**Comment by brianteeman** [comment]
> The instructions for this were helpful to me as well as the gif.

that would be the basis of the documentation

> Of course, I can't say whether I'll ever use it in conjunction with Editor.

100% not for everyone. It's something some 3pd editors offer and as we already shipped the library it made sense to make it available to those  that might want it.

File: administrator/language/en-GB/plg_editors_tinymce.ini [hunk 1]
```diff
@@ -50,6 +50,9 @@ PLG_TINY_FIELD_SKIN_ADMIN_LABEL="Administrator Skin"
 PLG_TINY_FIELD_SKIN_INFO_DESC="Copy your new skins to: /media/editors/tinymce/skins/ui."
 PLG_TINY_FIELD_SKIN_INFO_LABEL="For customised skins go to: <a href=\"http://skin.tiny.cloud\" target=\"_blank\">Skin Creator</a>"
 PLG_TINY_FIELD_SKIN_LABEL="Site Skin"
+;do not translate the word markdown
+PLG_TINY_FIELD_TEXTPATTERN_DESC="Support for basic Markdown syntax."
```
Comment (wilsonge) [inline administrator/language/en-GB/plg_editors_tinymce.ini]: Does the description here really add value?

Comment (infograf768) [reply administrator/language/en-GB/plg_editors_tinymce.ini]: It does not indeed. We need more as proposed above and boldly rejected.

Comment (brianteeman) [reply administrator/language/en-GB/plg_editors_tinymce.ini]: It is just as informative as the equivalent description in codemirror plugin

**Comment by toivo** [comment]
My 2c: this PR should be part of J4. Markdown is a welcome addition to the editing capabilities in Joomla, useful not only to those who have used editors at docs.joomla.org and wikipedia.org.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31911">issues.joomla.org/tracker/joomla-cms/31911</a>.</sub>

**Comment by toivo** [comment]
I have tested this item :white_check_mark: successfully on 5588789ed156d07ad756f27807031c959c20fd9d<br /><br />Tested successfully in Beta7-dev of 10 January.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31911">issues.joomla.org/tracker/joomla-cms/31911</a>.</sub>

**Comment by infograf768** [comment]
> My 2c: this PR should be part of J4. Markdown is a welcome addition to the editing capabilities in Joomla, useful not only to those who have used editors at docs.joomla.org and wikipedia.org.

Fine with that. I was wrong stating it can be considered as new feature. Nevertheless, we need a description with a real meaning/warning.

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on 5588789ed156d07ad756f27807031c959c20fd9d<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31911">issues.joomla.org/tracker/joomla-cms/31911</a>.</sub>

File: administrator/language/en-GB/plg_editors_tinymce.ini (hunk 1 above)
Comment (infograf768) [reply administrator/language/en-GB/plg_editors_tinymce.ini]: And that is why most basic users, if using codemirror, just use Default.
Vim is not even actively maintained.
https://codemirror.net/demo/vim.html

TinyMCE is the most used core J editor. We went as far as introducing various parameters depending on user status.

Comment (brianteeman) [reply administrator/language/en-GB/plg_editors_tinymce.ini]: that just confirmed everything I have said.
most users, if using tinymce, just use default

File: administrator/language/en-GB/plg_editors_tinymce.ini [hunk 2]
```diff
@@ -50,6 +50,9 @@ PLG_TINY_FIELD_SKIN_ADMIN_LABEL="Administrator Skin"
 PLG_TINY_FIELD_SKIN_INFO_DESC="Copy your new skins to: /media/editors/tinymce/skins/ui."
 PLG_TINY_FIELD_SKIN_INFO_LABEL="For customised skins go to: <a href=\"http://skin.tiny.cloud\" target=\"_blank\">Skin Creator</a>"
 PLG_TINY_FIELD_SKIN_LABEL="Site Skin"
+;do not translate the word markdown
+PLG_TINY_FIELD_TEXTPATTERN_DESC="Support for basic Markdown text formatting."
```
Comment (Quy) [inline administrator/language/en-GB/plg_editors_tinymce.ini]:
```suggestion
;do not translate the word Markdown
PLG_TINY_FIELD_TEXTPATTERN_DESC="Use Markdown syntax to compose content with links, lists, and other styles."
```

Comment (brianteeman) [reply administrator/language/en-GB/plg_editors_tinymce.ini]: If you know what markdown is - then the extra information is of no benefit
If you dont know what markdown is - then there extra information is not enough

Like a lot of optional settings in Joomla everyone can not be expected to know what the non-default settings are for. That's why we have the extensive help docs.

Comment (Quy) [reply administrator/language/en-GB/plg_editors_tinymce.ini]: I agree with you. Let's remove the description.

**Comment by wilsonge** [comment]
Let's go with it. We can update the description as required as we go

**Comment by brianteeman** [comment]
Thanks - I will update the docs site tomorrow

**Pull request merged**